File: src/main/java/org/c3lang/intellij/completion/TypeCompletionContributor.java
package org.c3lang.intellij.completion;

import com.intellij.codeInsight.completion.CompletionParameters;
import com.intellij.codeInsight.completion.CompletionProvider;
import com.intellij.codeInsight.completion.CompletionResultSet;
import com.intellij.codeInsight.completion.InsertHandler;
import com.intellij.codeInsight.completion.InsertionContext;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.patterns.ElementPattern;
import com.intellij.psi.PsiElement;
import com.intellij.psi.stubs.StubIndex;
import com.intellij.util.ProcessingContext;
import org.c3lang.intellij.C3Icons;
import org.c3lang.intellij.index.NameIndex;
import org.c3lang.intellij.intention.AddImportQuickFix;
import org.c3lang.intellij.project.C3ProjectService;
import org.c3lang.intellij.psi.C3FnParameterList;
import org.c3lang.intellij.psi.C3ModuleDefinition;
import org.c3lang.intellij.psi.C3PathIdent;
import org.c3lang.intellij.psi.C3PsiElement;
import org.c3lang.intellij.psi.C3Type;
import org.c3lang.intellij.psi.C3TypeFullyQualifiedNamePsiElement;
import org.c3lang.intellij.psi.C3TypeName;
import org.c3lang.intellij.psi.C3Types;
import org.c3lang.intellij.psi.FullyQualifiedName;
import org.c3lang.intellij.psi.ModuleName;
import org.c3lang.intellij.stubs.C3TypeEnum;
import org.jetbrains.annotations.NotNull;

import javax.swing.Icon;
import java.util.List;
import java.util.Objects;

import static com.intellij.patterns.PlatformPatterns.psiElement;
import static com.intellij.patterns.StandardPatterns.or;

@SuppressWarnings("DuplicatedCode")
public final class TypeCompletionContributor extends CompletionProvider<CompletionParameters>
{
    public static final TypeCompletionContributor INSTANCE = new TypeCompletionContributor();

    @SuppressWarnings("unused")
    private static final Logger LOG = Logger.getInstance(TypeCompletionContributor.class);

    private static final ElementPattern<PsiElement> PATTERN = or(
        psiElement().inside(C3PathIdent.class),
        psiElement(C3Types.TYPE_IDENT),
        psiElement().inside(C3FnParameterList.class)
    );

    private TypeCompletionContributor()
    {
    }

    @Override
    protected void addCompletions(
            @NotNull CompletionParameters parameters,
            @NotNull ProcessingContext context,
            @NotNull CompletionResultSet result)
    {
        if (!PATTERN.accepts(parameters.getPosition()) && !PATTERN.accepts(parameters.getOriginalPosition()))
        {
            return;
        }

        C3Type lookupTarget = CompletionExtensionsKt.siblingOf(parameters, C3Type.class);
        if (lookupTarget == null) return;

        String lookupString = CompletionExtensionsKt.getLookupString(parameters, lookupTarget);
        var matcher = CompletionExtensionsKt.getMatcher(lookupString);

        C3ModuleDefinition moduleDefinition = CompletionExtensionsKt.getModuleDefinition(parameters);
        if (moduleDefinition == null) return;

        var project = parameters.getPosition().getProject();
        ModuleName moduleName = moduleDefinition.getModuleName();
        InsertHandler<LookupElement> insertHandler = new StructInsertHandler(moduleDefinition, lookupTarget);

        for (String key : StubIndex.getInstance().getAllKeys(NameIndex.KEY, project))
        {
            if (!matcher.matches(key) && !key.isBlank()) continue;

            for (C3PsiElement element : StubIndex.getElements(
                    NameIndex.KEY,
                    key,
                    project,
                    C3ProjectService.getInstance(project).getSearchScope(),
                    C3PsiElement.class))
            {
                if (!(element instanceof C3TypeFullyQualifiedNamePsiElement typeName)) continue;

                Icon icon = iconFor(typeName.getTypeEnum());
                FullyQualifiedName fqName = typeName.getFqName();
                LookupElementBuilder lookupElementBuilder = LookupElementBuilder
                    .create(typeName, fqName.getFullName())
                    .withLookupStrings(List.of(
                        typeName.getFqName().getFullName(),
                        typeName.getFqName().getSuffixName(),
                        typeName.getFqName().getName()
                    ))
                    .withPsiElement(typeName)
                    .withIcon(icon)
                    .withPresentableText(Objects.equals(fqName.getModule(), moduleName)
                        ? fqName.getName()
                        : fqName.getFullName())
                    .withInsertHandler(insertHandler);

                result.addElement(lookupElementBuilder);
            }
        }
    }

    private static Icon iconFor(@NotNull C3TypeEnum typeEnum)
    {
        return switch (typeEnum)
        {
            case FALLBACK -> null;
            case STRUCT -> C3Icons.Nodes.STRUCT;
            case INTERFACE -> C3Icons.Nodes.INTERFACE;
            case ENUM -> C3Icons.Nodes.ENUM;
            case CONSTDEF -> C3Icons.Nodes.CONSTDEF;
            case UNION -> C3Icons.Nodes.UNION;
            case BITSTRUCT -> C3Icons.Nodes.BITSTRUCT;
            case FAULT -> C3Icons.Nodes.FAULT;
        };
    }

    @SuppressWarnings("DuplicatedCode")
    private static final class StructInsertHandler implements InsertHandler<LookupElement>
    {
        private final C3ModuleDefinition moduleDefinition;
        private final PsiElement lookupTarget;

        private StructInsertHandler(@NotNull C3ModuleDefinition moduleDefinition, @NotNull PsiElement lookupTarget)
        {
            this.moduleDefinition = moduleDefinition;
            this.lookupTarget = lookupTarget;
        }

        @Override
        public void handleInsert(@NotNull InsertionContext context, @NotNull LookupElement item)
        {
            PsiElement psiElement = item.getPsiElement();
            if (!(psiElement instanceof C3TypeName element)) return;

            WriteCommandAction.runWriteCommandAction(context.getProject(), () -> {
                AddImportQuickFix.ImportAction importAction =
                    AddImportQuickFix.addImportAsText(element, moduleDefinition);

                ModuleName importModuleName = importAction != null ? importAction.getModuleName() : null;
                String textToInsert = moduleDefinition.textToInsert(importModuleName, element);
                int endOffset = context.getEditor().getCaretModel().getOffset();

                context.getDocument().replaceString(
                    lookupTarget.getTextRange().getStartOffset(),
                    endOffset,
                    textToInsert
                );

                if (importAction != null)
                {
                    importAction.write(context.getDocument());
                }
            });
        }
    }
}

File: src/main/java/org/c3lang/intellij/completion/TypeCompletionContributor.kt
package org.c3lang.intellij.completion

import com.intellij.codeInsight.completion.*
import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.diagnostic.Logger
import com.intellij.patterns.PlatformPatterns.or
import com.intellij.patterns.PlatformPatterns.psiElement
import com.intellij.psi.PsiElement
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StubIndex
import com.intellij.refactoring.suggested.startOffset
import com.intellij.util.ProcessingContext
import org.c3lang.intellij.C3Icons
import org.c3lang.intellij.index.NameIndex
import org.c3lang.intellij.intention.AddImportQuickFix
import org.c3lang.intellij.psi.*
import org.c3lang.intellij.stubs.C3TypeEnum
import javax.swing.Icon

@Suppress("DuplicatedCode")
object TypeCompletionContributor : CompletionProvider<CompletionParameters>() {
    private val log = Logger.getInstance(
        TypeCompletionContributor::class.java
    )

    private val pattern = or(
        // foo::<caret>
        psiElement().inside(C3PathIdent::class.java),
        psiElement(C3Types.TYPE_IDENT),
        psiElement().inside(C3FnParameterList::class.java),
    )

    override fun addCompletions(
        parameters: CompletionParameters,
        context: ProcessingContext,
        result: CompletionResultSet
    ) {
        val originalPosition = parameters.originalPosition

        if (!pattern.accepts(parameters.position) && !pattern.accepts(parameters.originalPosition)) {
            return;
        }

        // TODO myData<caret>.open.result = ...;  should complete reference to local declaration instead of Type or maybe both
//        val elementPartOfReference = generateSequence(parameters.originalPosition ?: parameters.position) {
//            it.parent
//        }.any { it.reference?.resolve() != null }

        val lookupTarget = parameters.siblingOf<C3Type>()
            ?: return
        val lookupString = parameters.getLookupString(lookupTarget)
        val matcher = getMatcher(lookupString)

        val moduleDefinition = parameters.moduleDefinition ?: return
        val project = parameters.position.project
        val moduleName = moduleDefinition.moduleName

        val insertHandler = StructInsertHandler(
            moduleDefinition = moduleDefinition,
            lookupTarget = lookupTarget
        )

        StubIndex.getInstance().getAllKeys(
            NameIndex.KEY,
            project
        ).filter { matcher.matches(it) || it.isBlank() }.flatMap { key ->
            StubIndex.getElements(
                NameIndex.KEY,
                key,
                project,
                GlobalSearchScope.allScope(project),
                C3PsiElement::class.java,
            )
        }.filterIsInstance<C3TypeFullyQualifiedNamePsiElement>().forEach { typeName ->
            val icon: Icon? = when (typeName.typeEnum) {
                C3TypeEnum.FALLBACK -> null
                C3TypeEnum.STRUCT -> C3Icons.Nodes.STRUCT
                C3TypeEnum.INTERFACE -> C3Icons.Nodes.INTERFACE
                C3TypeEnum.ENUM -> C3Icons.Nodes.ENUM
                C3TypeEnum.UNION -> C3Icons.Nodes.UNION
                C3TypeEnum.BITSTRUCT -> C3Icons.Nodes.BITSTRUCT
                C3TypeEnum.FAULT -> C3Icons.Nodes.FAULT
            }

            val fqName = typeName.fqName
            val lookupElementBuilder = LookupElementBuilder
                .create(typeName, fqName.fullName)
                .withLookupStrings(
                    listOf(
                        typeName.fqName.fullName,
                        typeName.fqName.suffixName,
                        typeName.fqName.name
                    )
                )
                .withPsiElement(typeName)
                .withIcon(icon)
                .withPresentableText(if (fqName.module == moduleName) fqName.name else fqName.fullName)
                .withInsertHandler(insertHandler)

            result.addElement(lookupElementBuilder)
        }
    }

    @Suppress("DuplicatedCode")
    private class StructInsertHandler(
        private val moduleDefinition: C3ModuleDefinition,
        private val lookupTarget: PsiElement,
    ) : InsertHandler<LookupElement> {

        override fun handleInsert(context: InsertionContext, item: LookupElement) {
            val editor = context.editor
            val document = editor.document
            val element = item.psiElement as? C3TypeName
                ?: return

            WriteCommandAction.runWriteCommandAction(context.project) {
                val importAction = AddImportQuickFix.addImportAsText(
                    element = element,
                    moduleSection = moduleDefinition,
                )

                val textToInsert = moduleDefinition.textToInsert(importAction?.moduleName, element)
                val endOffset = editor.caretModel.offset

                document.replaceString(
                    lookupTarget.textRange.startOffset,
                    endOffset,
                    textToInsert
                )

                importAction?.write(document)

//                PsiDocumentManager.getInstance(context.project).commitDocument(document)
            }
        }

    }

}
